File: PGS/Part05/JaeHyeon/L2_87946.java
class Solution {
    static int answer = 0;
    public int solution(int k, int[][] dungeons) {
    
        dfs(k, dungeons, new boolean[dungeons.length], 0);
        return answer;
    }
    
    void dfs(int k, int[][] dungeons, boolean[] visited, int cnt) {
        
        boolean canExplorer = false;
        
        for (int i = 0; i < visited.length; i++) {
            if (!visited[i]) {
                if (dungeons[i][0] <= k) {
                    canExplorer = true;
                    break;
                }            
            }
        }
        
        if (!canExplorer) {
            answer = Math.max(answer, cnt);
            return;
        }
        
        for (int i = 0; i < visited.length; i++) {
            if (!visited[i]) {
                if (dungeons[i][0] <= k) {
                    visited[i] = true;
                    dfs(k - dungeons[i][1], dungeons, visited, cnt + 1);
                    visited[i] = false;
                }
            }
        }
    } 
}

/*
최소 필요 피로도, 소모 피로도

*/

File: PGS/Part05/SoHyun/L2_87946.py
answer = 0

def dfs(k, cnt, dungeons, v):
    global answer
    if cnt > answer:
        answer = cnt
        
    for i in range(len(dungeons)):
        if not v[i] and k >= dungeons[i][0]:
            v[i] = 1
            dfs(k-dungeons[i][1], cnt+1, dungeons, v)
            v[i] = 0

def solution(k, dungeons):
    global answer
    v = [0] * len(dungeons)
    dfs(k, 0, dungeons, v)
    return answer